pkg/storage/ingest/pusher.go

```diff
@@ -75,14 +79,32 @@ func (c pusherConsumer) pushRequests(ctx context.Context, reqC <-chan parsedReco
 
 		c.processingTimeSeconds.Observe(time.Since(processingStart).Seconds())
 		if err != nil {
-			level.Error(c.l).Log("msg", "failed to push write request; skipping", "err", err)
-			// TODO move distributor's isClientError to a separate package and use that here to swallow only client errors and abort on others
-			continue
+			if !isClientIngesterError(err) {
+				return fmt.Errorf("consuming record at index %d for tenant %s: %w", recordIdx, wr.tenantID, err)
+			}
+			level.Warn(c.l).Log("msg", "consuming write request", "err", err, "user", wr.tenantID)
 		}
 	}
 	return nil
 }
 
+func isClientIngesterError(err error) bool {
+	stat, ok := grpcutil.ErrorToStatus(err)
+	if !ok {
+		// This should not be reached but in case it is, fall back to assuming it's our fault.
+		return false
+	}
+
+	if details := stat.Details(); len(details) > 0 {
+		if errDetails, ok := details[0].(*mimirpb.ErrorDetails); ok {
+			// This is usually the case.
+			return errDetails.Cause == mimirpb.BAD_DATA
```

Comment:
This means that `INSTANCE_LIMIT` will be retried, but will keep failing. I'm wondering how `INSTANCE_LIMIT` could work with ingest storage.

For example, today we only block creation of new series when `errMaxInMemorySeriesReached` is reached, but after this change we'll keep retrying ingesting the same request over and over. The request could be partially ingested tho.

Reply:
i think the problem comes from the fact that we can't rely on the client to retry in this case. Today we leave the responsibility of retrying 5xx to the client prometheus and don't block ingestion of other clients if they don't hit this instance limit. But this is more complicated with kafka. If we ignore the errors and continue ingesting series which don't trigger the instance limit, then we incur data loss.

One way to achieve the same behaviour with kafka is to have a dead-letter queue, where we push requests we couldn't process because of instance limits (or other errors). But I think this comes with a lot of complications: ordering, timely consumption from the DLQ, and deciding when to consume from the DLQ for example.

IMO the way to recover from this instance error is to wait for the series in the head block to fall out of retention and be truncated or manually scale the resources and raise the instance limit.

Reply:
I agree with you. Let's keep thinking whether we can improve the behaviour on instance limits, but let's not block on this.